### ISSUE TITLE
Must some `-Ctarget-feature`s be restrained on RISCV targets?

### ISSUE THREAD

**Issue by workingjubilee** [rust-lang/rust#132618]
refs:
- https://github.com/rust-lang/rust/issues/116344
- https://github.com/rust-lang/rust/pull/129884
- https://github.com/rust-lang/rust/issues/131799
- https://github.com/rust-lang/rust/pull/131807

in discussion of the target modifiers RFC[^1], Ralf asks this question:

> That's actually fine on most targets, e.g. on x86 if you set the soft-float target feature then you can enable x87 and SSE and whatnot and keep using the soft-float ABI. Only aarch64 https://github.com/llvm/llvm-project/issues/110632 (from the targets I checked so far).
>
> The interesting question is what happens on RISC-V when I disable the FP instructions on a target that usually uses the hardfloat ABI. Sadly, what LLVM usually does in that case is silently fall back to the softfloat ABI. But I haven't checked for RISC-V.

cc @beetrees 

[^1]: https://github.com/rust-lang/rfcs/pull/3716

**Comment by workingjubilee** [comment]
cc @coastalwhite

**Comment by workingjubilee** [comment]
Thanks for the summary! That seems pretty definite, yeah.

**Comment by RalfJung** [comment]
> Targets using any ABI other than ilp32e/lp64e rely on the e feature being disabled.

Wait so *enabling* `e` actually makes *fewer* things work? 😢 and here I was thinking RISC-V actually was a sane target... the float ABI stuff in RISC-V actually makes sense unlike in any other target I looked at so far, but having negative features is a nightmare.

**Comment by workingjubilee** [comment]
> > Targets using any ABI other than ilp32e/lp64e rely on the e feature being disabled.
> 
> Wait so _enabling_ `e` actually makes _fewer_ things work? 😢 and here I was thinking RISC-V actually was a sane target... the float ABI stuff in RISC-V actually makes sense unlike in any other target I looked at so far, but having negative features is a nightmare.

technically the way it actually works is that the architectures are, from the high-level perspective,
- `rv32i`
- `rv64i`
- `rv128i`(?)
- `rv32e`
- `rv64e`

**Comment by workingjubilee** [comment]
relevant to this in a slightly sideways manner: https://github.com/rust-lang/rust/pull/134358 will expose a way to `cfg` on RV32E targets without requiring a specific disposition for which way we zig or zag between `-e` or `+i`

**Comment by RalfJung** [comment]
> Looking at the RISC-V ISA specification, the RV32E/RV64E base instruction sets are described as reduced versions of RV32I/RV64I (the only difference being fewer available GPR registers): this seems similar to the way that e.g. the Zfhmin extension is a subset of the Zfh extension. So rather than having `e` as a negative feature, I think it would make more sense to have an `i` feature which is enabled when targeting RV32I/RV64I and disabled when targeting RV32E/RV64E. This allow things like `#[target_feature(enable = "i")]` to work as expected, and would be much more consistent.

Yeah that makes sense, but it would divert from how RISCV target features are handled in other languages. Seems like a discussion for https://github.com/rust-lang/rust/issues/114544.

> i.e. LLVM representing it as a feature is on LLVM, and from the RISCV architectural definition perspective these are all separate "base" architectures.

Oh wait this is an LLVM thing after all? What does GCC do?

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/134337 catches most of the potential ABI issues listed above. What remains are problems caused by feature implications that rustc does not know about. I have some plans here but that will have to wait for a future PR.

**Comment by workingjubilee** [comment]
I'm not sure gcc doesn't do the same thing, I'm addressing "how the specification for RISCV itself reasons about the underlying thing going on here". Essentially, the answer from the hardware architects might be that there are really 4 different `target_arch` values:
- "riscv32i"
- "riscv64i"
- "riscv32e"
- "riscv64e"

It seems doubtful we would want to actually go that route, but nonetheless!

**Comment by RalfJung** [comment]
So... is it even worth supporting toggling the "e" feature then? We don't support toggling the target architecture, either...

**Comment by workingjubilee** [comment]
@RalfJung Doubtful, and if we did, it might be in the direction that beetrees mentioned.

**Comment by RalfJung** [comment]
Are there any other features besides `d` that are incompatible with `ilp32e` due to stack alignment issues (or other concerns)?

**Comment by workingjubilee** [comment]
...hmmm.

...RISCV vectors must be accessed at the alignment of their element, which is configured by the Selected Element Width value. This suggests we cannot use vectors with 64-bit alignment on ilp32e either? There's some RISCV features that depend on interacting with 64-bit elements, so we should probably assume they're all no-go until proven otherwise?

cc @topperc do I understand that correctly?

**Comment by topperc** [comment]
I'm not sure LLVM still has a fatal error for ilp32e+D. The default stack alignment for ilp32e is 4, but LLVM should be capable of realigning the stack if needed.

**Comment by workingjubilee** [comment]
Ah yeah, I guess that's not a concern anymore.

**Comment by RalfJung** [comment]
So  should we remove that error from our end as well then? If so, please file an MR. :)

**Comment by workingjubilee** [comment]
it might be LLVM version dependent, in which case we might need to keep our error around until we drop the last version that errored on it.

**Comment by RalfJung** [comment]
The current ABI checking logic for RISCV looks like this:
https://github.com/rust-lang/rust/blob/1dc023e9da6038d0be2e3ec8b25f15ea59a83428/compiler/rustc_target/src/target_features.rs#L1236-L1269

This takes into account feature implications and `-Ctarget-cpu`.
I think that means we are done here? Or is there anything else left to check?
